packages/nextcloud/lib/src/webdav/path_uri.dart

```diff
@@ -0,0 +1,100 @@
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+
+/// A [Uri] that only exposes the path handling.
+@immutable
+class PathUri {
+  /// Creates a new path URI.
+  ///
+  /// Only one of [path] and [pathSegments] must be given.
+  /// All duplicate slashes are automatically removed.
+  /// An empty path will be considered as the current working directory.
+  PathUri({
+    final String? path,
+    final List<String>? pathSegments,
+  })  : assert(
+          (path == null) != (pathSegments == null) || (path == null && pathSegments == null),
+          'Only one of path and pathSegments must be given',
+        ),
+        _segments = _buildSegments(
+          path: path,
+          pathSegments: pathSegments,
+        );
+
+  static List<String> _buildSegments({
+    final String? path,
+    final List<String>? pathSegments,
+  }) {
+    final segments = pathSegments ?? path?.split('/');
+    if (segments == null) {
+      return [];
+    }
+    return segments
+        .whereIndexed(
+          (final i, final s) => i == 0 || i == segments.length - 1 || s.isNotEmpty,
+        )
+        .toList();
+  }
+
+  final List<String> _segments;
+
+  List<String> get _nonEmptySegments => _segments.where((final s) => s.isNotEmpty).toList();
+
+  /// Returns the path as a string.
+  ///
+  /// See [pathSegments] for getting the path as a list of its segments.
+  String get path => _segments.join('/');
+
+  /// Returns the path as a list of its segments.
+  ///
+  /// See [path] for getting the path as a string.
+  List<String> get pathSegments => _segments;
+
+  /// Returns whether the path is a directory.
+  ///
+  /// If the whole path is empty it is considered to be a directory.
+  bool get isDirectory => _segments.lastOrNull?.isEmpty ?? true;
+
+  /// Returns the name of the last element in path.
+  String get name => _nonEmptySegments.lastOrNull ?? '';
+
+  /// Returns the parent of the path.
+  PathUri? get parent {
+    final nonEmptySegments = _nonEmptySegments;
+    if (nonEmptySegments.length > 1) {
+      return PathUri(
+        pathSegments: [
+          // Potential leading slash
+          if (_segments.first.isEmpty) '',
+          ...nonEmptySegments.sublist(0, nonEmptySegments.length - 1),
+          // Accessing the parent will always lead to a directory so we always have a trailing slash
+          '',
```

Comment:
I have not compleately wrapped my head arround it but I don't like the empty strings in here and somehow don't really understand why we have them.

If this is solely to get a a leading `/` then there are better ways to do this. Maybe the dart `path` library?

Reply:
We could store this in new fields `isAbsolute` and `isDirectory`, but then it becomes more annoying and error prone to modify and merge paths. It would be a lot cleaner though so we should probably do it. As long as we set both to be required in the constructor we shouldn't make to many mistakes.